### PR TITLE
WIP: BLD: prepend numpy include dirs

### LINKED ISSUE #11853: BLD: unable to build from source over existing installation in non-standard path
Dear numpy developers,

I have a CentOS 7 machine with python3.6 installed from https://www.softwarecollections.org/en/scls/rhscl/rh-python36/

This installs python3 into the non-standard path
```
/opt/rh/rh-python36/root/usr
```

The collection also comes with a numpy 1.13, installed in the same path.
As I do not have root access to the machine, I would like to upgrade to the current development version building the master version of numpy from source and installing it in a user folder.

The system folder
```
/opt/rh/rh-python36/root/usr/include
```
contains python3.6m includes, but also a symbolic link
```
numpy -> /opt/rh/rh-python36/root/usr/lib64/python3.6/site-packages/numpy/core/include/numpy/
```
This causes the compilation of the multiarray package (see [1] for the full command)
```
gcc .... -I/opt/rh/rh-python36/root/usr/include .... -Ibuild/src.linux-x86_64-3.6/numpy/core/src/private -Inumpy/core/include ...
```
to include non-standard system-wide headers (from /opt/rh/rh-python36/root/usr/include/numpy, v 1.13) before the master ones, resulting in several occurences of
```
error: ‘NPY_ARRAY_WRITEBACKIFCOPY’ undeclared (first use in this function)
```
because NPY_ARRAY_WRITEBACKIFCOPY was defined in numpy >= 1.14.

Is there a way to configure the setup script to move the
```
-I/opt/rh/rh-python36/root/usr/include
```
after all others include flags? 

On an unrelated note, is the
```
/opt/rh/rh-python36/root/usr/include/numpy
```
symbolic link really needed for numpy to work properly?

Thanks,
best regards,
Francesco Ballarin

[1] The full command is 
```
gcc -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv  -I/opt/rh/rh-python36/root/usr/include -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DSCIPY_MKL_H -DHAVE_CBLAS -Ibuild/src.linux-x86_64-3.6/numpy/core/src/private -Inumpy/core/include -Ibuild/src.linux-x86_64-3.6/numpy/core/include/numpy -Inumpy/core/src/private -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/opt/rh/rh-python36/root/usr/include/python3.6m -Ibuild/src.linux-x86_64-3.6/numpy/core/src/private -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/private -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/private -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath  -c numpy/core/src/multiarray/compiled_base.c -o build/temp.linux-x86_64-3.6/numpy/core/src/multiarray/compiled_base.o -MMD -MF build/temp.linux-x86_64-3.6/numpy/core/src/multiarray/compiled_base.o.d
```

### PR DESCRIPTION
**Pull request by mattip**
Fixes #11853 by prepending the `numpy.distutil.misc_util.get_numpy_include_dirs()` to the `build_c_lib` and `build_extension` `include_dirs`.

Added tests that fail before, pass after the  `distutil` changes. Is there a case where using our includes will break downstream packages?

@rainwoodman , @francesco-ballarin, could you test this by either cloning my branch or making the changes in `build_clib.py` and `build_ext.py` locally?

Edit: add WIP label, need to figure out what lib path is missing

### CODE REVIEW

**Comment by mattip** [comment]
It seems my `Extension` mocking is incomplete. Added WIP tag until tests are passing.

File: doc/release/1.16.0-notes.rst [hunk 1]
```diff
@@ -148,6 +148,12 @@ behavior will be appending.  This applied to: `LDFLAGS`, `F77FLAGS`,
 `F90FLAGS`, `FREEFLAGS`, `FOPT`, `FDEBUG`, and `FFLAGS`.  See gh-11525 for more
 details.
 
+Prefer our include files
+------------------------
+We now prepend the `include/numpy` directory to the `include_dirs` when using
+`nupy.distutils`'s `build_ext` and `build_clib` commands. Previously we appended
+our include dir, which sometimes caused problems when building NumPy.
```
Comment (eric-wieser) [inline doc/release/1.16.0-notes.rst]: Better as "this include directory" rather that "our include dir", to make it clear it's referring to the same one

**Comment by rainwoodman** [comment]
This still fails building with Anaconda's cross compilation toolchain. Looking at the log, it is because /usr/include exists in the compilation flags at all, rather than the location of it.

The term is found by numpy due to existing librarys matched under /usr/ prefix. However, under the cross compilation toolchain files under /usr are not for the target platform, and shall not be checked even if they can be linked.

To solve this, I think instead of a hard coded /usr prefix, numpy shall look for libraries from the sys prefix reported by the C compiler, e.g.

```
$ /home/yfeng1/anaconda3/install/bin/x86_64-conda_cos6-linux-gnu-cc -print-sysroot
/home/yfeng1/anaconda3/install/bin/../x86_64-conda_cos6-linux-gnu/sysroot
```

When not cross compiling, the command may return an empty string
```
$ gcc -print-sysroot

```

Here is the current error:

```
/home/yfeng1/anaconda3/install/bin/x86_64-conda_cos6-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -pipe -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -fPIC -DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DSCIPY_MKL_H -DHAVE_CBLAS -Inumpy/core/include -Ibuild/src.linux-x86_64-3.6/numpy/core/include/numpy -I/usr/local/include -I/usr/include -I/usr/include/suitesparse -I/home/yfeng1/anaconda3/install/include -Ibuild/src.linux-x86_64-3.6/numpy/core/src/umath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/home/yfeng1/anaconda3/install/include/python3.6m -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -c numpy/core/src/multiarray/numpyos.c -o build/temp.linux-x86_64-3.6/numpy/core/src/multiarray/numpyos.o -MMD -MF build/temp.linux-x86_64-3.6/numpy/core/src/multiarray/numpyos.o.d
In file included from numpy/core/src/multiarray/numpyos.c:23:0:
/home/yfeng1/anaconda3/install/x86_64-conda_cos6-linux-gnu/sysroot/usr/include/xlocale.h:28:16: error: redefinition of 'struct __locale_struct'
 typedef struct __locale_struct
                ^~~~~~~~~~~~~~~
In file included from /usr/include/bits/types/locale_t.h:22:0,
                 from /usr/include/string.h:152,
                 from /home/yfeng1/anaconda3/install/include/python3.6m/Python.h:30,
                 from numpy/core/src/multiarray/numpyos.c:2:
/usr/include/bits/types/__locale_t.h:28:8: note: originally defined here
 struct __locale_struct
        ^~~~~~~~~~~~~~~
In file included from numpy/core/src/multiarray/numpyos.c:23:0:
/home/yfeng1/anaconda3/install/x86_64-conda_cos6-linux-gnu/sysroot/usr/include/xlocale.h:40:4: error: conflicting types for '__locale_t'
 } *__locale_t;
    ^~~~~~~~~~
In file included from /usr/include/bits/types/locale_t.h:22:0,
                 from /usr/include/string.h:152,
                 from /home/yfeng1/anaconda3/install/include/python3.6m/Python.h:30,
                 from numpy/core/src/multiarray/numpyos.c:2:
/usr/include/bits/types/__locale_t.h:42:33: note: previous declaration of '__locale_t' was here
 typedef struct __locale_struct *__locale_t;
                                 ^~~~~~~~~~
In file included from numpy/core/src/multiarray/numpyos.c:23:0:
/home/yfeng1/anaconda3/install/x86_64-conda_cos6-linux-gnu/sysroot/usr/include/xlocale.h:43:20: error: conflicting types for 'locale_t'
 typedef __locale_t locale_t;
                    ^~~~~~~~
In file included from /usr/include/string.h:152:0,
                 from /home/yfeng1/anaconda3/install/include/python3.6m/Python.h:30,
                 from numpy/core/src/multiarray/numpyos.c:2:
/usr/include/bits/types/locale_t.h:24:20: note: previous declaration of 'locale_t' was here
 typedef __locale_t locale_t;
                    ^~~~~~~~
numpy/core/src/multiarray/numpyos.c: In function 'NumPyOS_ascii_strtold':
numpy/core/src/multiarray/numpyos.c:568:13: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
     clocale = newlocale(LC_ALL_MASK, "C", NULL);
             ^
numpy/core/src/multiarray/numpyos.c:571:39: warning: passing argument 3 of 'strtold_l' from incompatible pointer type [-Wincompatible-pointer-types]
         result = strtold_l(s, endptr, clocale);
                                       ^~~~~~~
In file included from /home/yfeng1/anaconda3/install/include/python3.6m/Python.h:34:0,
                 from numpy/core/src/multiarray/numpyos.c:2:
/usr/include/stdlib.h:230:20: note: expected 'locale_t {aka struct __locale_struct *}' but argument is of type 'locale_t {aka struct __locale_struct *}'
 extern long double strtold_l (const char *__restrict __nptr,
                    ^~~~~~~~~
numpy/core/src/multiarray/numpyos.c:572:20: warning: passing argument 1 of 'freelocale' from incompatible pointer type [-Wincompatible-pointer-types]
         freelocale(clocale);
                    ^~~~~~~
In file included from numpy/core/src/multiarray/numpyos.c:4:0:
/usr/include/locale.h:180:13: note: expected 'locale_t {aka struct __locale_struct *}' but argument is of type 'locale_t {aka struct __locale_struct *}'
 extern void freelocale (locale_t __dataset) __THROW;
             ^~~~~~~~~~
```

**Comment by rainwoodman** [comment]
I was able to compile after this change:

```
diff --git a/numpy/distutils/system_info.py b/numpy/distutils/system_info.py
index 79adcc334..f16098a25 100644
--- a/numpy/distutils/system_info.py
+++ b/numpy/distutils/system_info.py
@@ -262,11 +262,11 @@ if sys.platform == 'win32':
             add_system_root(os.path.join(conda_dir, 'Library'))
 
 else:
-    default_lib_dirs = libpaths(['/usr/local/lib', '/opt/lib', '/usr/lib',
+    default_lib_dirs = libpaths(['/usr/local/lib', '/opt/lib',
                                  '/opt/local/lib', '/sw/lib'], platform_bits)
     default_runtime_dirs = []
     default_include_dirs = ['/usr/local/include',
-                            '/opt/include', '/usr/include',
+                            '/opt/include',
                             # path of umfpack under macports
                             '/opt/local/include/ufsparse',
                             '/opt/local/include', '/sw/include',
```

I am not yet sure what to do about those X11 search paths; I think they will cause problems cross compiling other packages that uses numpy's distutils down the line.

I wonder if @mingwandroid can comment on the thread..

**Comment by mingwandroid** [comment]
Well, your patch is pretty much a reasonable fix, albeit one that's not appropriate for non-Anaconda builds (and so would have to be made optional, and I'd go further and disable *all* of those additions, we simply do not need them (and in fact most compilers do not need most of them!).

I think numpy developers may be best placed to comment on this, but first, pinging @msarahan who deals with numpy more frequently than I.

**Comment by rgommers** [comment]
> I am not yet sure what to do about those X11 search paths; I think they will cause problems cross compiling other packages that uses numpy's distutils down the line.

Those are not needed for numpy nor for any other package I'm familiar with. They were put in ages ago to support <some random packages>. They shouldn't cause any problems unless someone is actually using X11.

**Comment by rainwoodman** [comment]
X11 are probably needed for those gdk stuff, which no one uses anyways? Even if we are keeping them, on a modern system where gdk are declared in pkgconfig, X11 can probably be determined from pkgconfig too. Also, what about wayland?

I also noticed that paths under sys.prefix are already automatically added to the defaults a few lines later. So at least these /usr paths can be safely removed; these sys.prefix lines will automatically be replaced by the paths Python thinks are appropriate -- though likely still not really cross-compilation friendly. (Python is ran on the compiling host, not the compiled target)

/sw, /opt and /usr/local paths are relevant on some old sites and homebrew macs; no harm keeping them.

**Comment by mingwandroid** [comment]
> /sw, /opt and /usr/local paths are relevant on some old sites and homebrew macs; no harm keeping them.

Any include directories that contain headers that can conflict (on anyone's system) with the actual ones you want should be removed ideally both in general but much more importantly in the case of using pseudo-cross compilers like we do.

We have our own sysrooted headers that the compiler already knows where to find. I am somewhat tempted to 'poison' `/usr` from our compilers and linkers so nothing gets found in there ever.

**Comment by mattip** [comment]
@rainwoodman just to be clear

>  I was able to compile after this change:

You didn't use the changes I did in `numpy.distutils`, right?

**Comment by rainwoodman** [comment]
I made the change based on your branch.

I think your change is still logically correct and it may solve @francesco-ballarin 's case.

My case has an additional problem where the include list is not empty when it should.

**Comment by rainwoodman** [comment]
> > /sw, /opt and /usr/local paths are relevant on some old sites and homebrew macs; no harm keeping them.
> 
> Any include directories that contain headers that can conflict (on anyone's system) with the actual ones you want should be removed ideally both in general but much more importantly in the case of using pseudo-cross compilers like we do.

I think those heuristics are in distutils such that one does not need to modify site.cfg on those platforms. It must have helped early days porting and deploying numpy; but it does create this situation where the header and library used are somewhat randomly decided by the machine, not by the person.

> 
> We have our own sysrooted headers that the compiler already knows where to find. I am somewhat tempted to 'poison' `/usr` from our compilers and linkers so nothing gets found in there ever.

That indeed would be a practical solution to this issue without requiring additional changes to numpy, though it feels like patching over patches.

@mattip Shall I file a different PR for the removal of /usr  and default paths?

**Comment by mattip** [comment]
> Shall I file a different PR for the removal of /usr and default paths?

@rainwoodman yes please, although I fear trying to mock and write tests for those changes. I can't even get the current simple change's tests to pass.

**Comment by rainwoodman** [comment]
I guess I'll have to see how far we can go with the test here. It won't decrease the test coverage as it is removing lines, and any test we write won't be able to cover all real world use cases.

Did you see if the tests in python's distutils has bits that can be reused here?

https://github.com/python/cpython/blob/master/Lib/distutils/tests/test_build_ext.py

**Comment by francesco-ballarin** [comment]
@mattip Not sure why, but it is still not working on my system.

```
error: Command "gcc -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -I/opt/rh/rh-python36/root/usr/include -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -Inumpy/core/include -Ibuild/src.linux-x86_64-3.6/numpy/core/include/numpy -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/opt/rh/rh-python36/root/usr/include/python3.6m -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -c build/src.linux-x86_64-3.6/numpy/core/src/multiarray/_multiarray_tests.c -o build/temp.linux-x86_64-3.6/build/src.linux-x86_64-3.6/numpy/core/src/multiarray/_multiarray_tests.o -MMD -MF build/temp.linux-x86_64-3.6/build/src.linux-x86_64-3.6/numpy/core/src/multiarray/_multiarray_tests.o.d" failed with exit status 1
```
with the same errors as before. The include 
```
-I/opt/rh/rh-python36/root/usr/include
'''
is still prepended to the others.

Remote is correctly set to https://github.com/mattip/numpy.git on include_dir-order branch.

Thanks.

**Comment by mattip** [comment]
@francesco-ballarin can you set a breakpoint in `numpy/distutils/command/build_ext.py` in the `build_extension` function and try to work out what is going on in the call to `self.compiler.compile`?

**Comment by rainwoodman** [comment]
- Another useful place to look is run `setup.py config`. It shows which dependency pulled in which path, if you parse it carefully. 
- Also may be useful is to see if the path is in any of you environment variables? 
- Do you have a site.cfg?

**Comment by francesco-ballarin** [comment]
Thanks, I will test these suggestions on Monday.

**Comment by francesco-ballarin** [comment]
I managed to debug a bit, to find that the string with the wrong include is stored in UnixCCompiler.compiler_so when printing it out in numpy/distutils/unixccompiler.py, after line 24. Unfortunately I have no idea where this compiler_so attribute is set, and a search with grep did not help me in finding it. Any suggestion?

I do have a site.cfg (to use intel mkl), but the same behavior is there using the default site.cfg.example. I do have environment variables set (most notable, INCLUDE and CPATH), but none of them is referring to the specific include directory which gets added to compiler_so.

Thanks.

**Comment by mattip** [comment]
@francesco-ballarin `compiler_so` appears in various places in the stdlib `distutils`. I think it is set in `distutils/sysconfig.py` in `customize_compiler`. In any case, this branch does not help solve the issue it was meant to solve, so I think it should be closed.

**Comment by francesco-ballarin** [comment]
I agree. Thanks for the work on this, looking forward to test other possible patches. (maybe one as simple as replacing -I with -isystem in compiler_so, or some better way to achieve the same goal?)

**Pull request closed without merging**